### DISCUSSION TITLE
Using @QuarkusMain in Kotlin produces warnings on startup

### DISCUSSION THREAD

**Discussion by jehrenzweig-leagueapps** [quarkusio/quarkus#26271] [Q&A] [answered]
I'm following the "_Application Initialization and Termination_" guide's instructions on [how to add a custom startup method to a Quarkus application](https://quarkus.io/guides/lifecycle#the-main-method).  The guide is written for Java but I'm using Kotlin; here's the code I've written thus far:
```
package com.leagueapps.microservices.poc.potato.application

import io.quarkus.runtime.Quarkus
import io.quarkus.runtime.annotations.QuarkusMain

@QuarkusMain
object ApplicationStartup {
	@JvmStatic
	fun main(args: Array<String>) {
		Quarkus.run(*args)
	}
}
```

The following item now appears in my log when the Quarkus application starts up:
```
2022-06-21 14:26:00,973 INFO  [io.qua.arc.pro.BeanProcessor] (build-31) Found unrecommended usage of private members (use package-private instead) in application beans:
	- Bean constructor void <init>() on com.leagueapps.microservices.poc.potato.application.ApplicationStartup
```

Are there any changes I can/should make to my code?  The message doesn't really give me much to go on... my code doesn't declare any private member, and as far as I know Kotlin doesn't support making anything package-private through the use of visibility keywords.  I also don't know what the `Bean constructor void <init> on` message means.

[a comment by an automated account is not shown]

**Comment by jehrenzweig-leagueapps** [comment, marked as answer]
Through the power of trial and error, I found a Kotlin code solution that does not produce the warning:
```
import io.quarkus.runtime.Quarkus
import io.quarkus.runtime.annotations.QuarkusMain

@QuarkusMain
class ApplicationStartup {
	companion object {
		@JvmStatic
		fun main(args: Array<String>) {
			Quarkus.run(*args)
		}
	}
}
```


And this seems to be a way to declare a code block like the second one mentioned in the guide, but in Kotlin:
```
import io.quarkus.runtime.Quarkus
import io.quarkus.runtime.QuarkusApplication
import io.quarkus.runtime.annotations.QuarkusMain

@QuarkusMain
class ApplicationStartup {
	companion object {
		@JvmStatic
		fun main(args: Array<String>) {
			println("Running ApplicationStartup.main() method")
			Quarkus.run(MyApp::class.java, *args)
		}
	}

	class MyApp : QuarkusApplication {
		override fun run(vararg args: String): Int {
			println("Do startup logic here")
			Quarkus.waitForExit()
			return 0
		}
	}
}
```

**Comment by evanchooly** [comment]
In your first attempt, you were using a top level `object`. These are singleton by definition and uninstantiable because the constructor is private.  your second attempt used an actual class whose default/noarg constructor is public by default.  So when the static main is attached to that class, there's a public constructor.  When it's attached to that `object`, there is not.

**Reply by jehrenzweig-leagueapps** [reply]
Great, thank you for that info.  I figured there was a reasonable explanation for the behavior... I just wasn't finding it, & didn't know how to go about searching for it.  Still learning the ropes w/ both Kotlin & Quarkus.
